enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,441 @@
+  operation done by the user is only on the `.metadata.annotations`.
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-name`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+Since the secrets are the primary resource for the secret-injector controller, the new controller
+will also set up watches and re-sync routes in the same namespace as that of the secret (similar
+to ingress-to-route controller).
+
+The new controller will watch all routes and process all routes in order to update
+`.status.certificate`. The ones with the annotation will have additional
+logic to update `.spec.tls.certificate` and `.spec.tls.key` based on the following pre-conditions,
+
+- The secret-reference annotation is present.
+- Secret mentioned in the annotation exists and is valid (contains the required fields mentioned [here](#api-extensions))
+
+In scenarios where the user deletes the annotation, it is the controller's
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e.
+the latest certificate status condition has to be `ManagedCertificate` and since the required annotation is
+not present this results in the `.spec.tls.certificate` and `.spec.tls.key` being cleared
+out and the `.status` updated to `DefaultCertificate`.
+
+If the user deletes the secret associated with the Route without deleting the
+`route.openshift.io/tls-secret-name` annotation, the secret-injector controller will
+clear out `.spec.tls.certificate` and `.spec.tls.key` since the following pre-conditions
+are met,
+
+- The secret-reference annotation is present.
+- Secret mentioned in the annotation doesn't exist
+- The latest condition on Route's `.status.certificate` has `ManagedCertificate`=`True`
+
+This results in the router using the default certificates that are generated and
+the `.status` updated as well.
+
+As a follow up to the above scenario when the secret-reference annotation added by the user is retained
+by the secret-injector controller and because the referenced secret is not present,
+the controller will publish an `Event` so that the end user is notified regarding this
+switch to using the default generated certificates. The `.status` is also updated to `DefaultCertificate`.
+
+When transitioning from using a managed certificate to manually managed certificate, the user
+is expected to delete the secret-reference annotation and provide new certificate/key pair under `.spec.tls`.
+If the previously used `.spec.tls.certificate` and `.spec.tls.key` is not replaced
+by the user, the secret-injector controller will clear `.spec.tls.certificate` and `.spec.tls.key`
```

Comment:
How will the controller know if it was the value from the secret or the new values from the user? If this were a spec field referencing a secret, this would be unambiguous